keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+
+```go
+type Provider struct {
+    // Issuer is a basic OIDC provider connection options.
+    Issuer Issuer
+    // ClaimValidationRules are rules that are applied to validate ID token claims to authorize users.
+    // +optional
+    ClaimValidationRules []ClaimValidationRule
+    // ClaimMappings points claims of an ID token to be treated as user attributes.
+    // All mappings are logical expressions that is written in CEL https://github.com/google/cel-go.
+    ClaimMappings UserAttributes
+    // ClaimsFilter allows unmarshalling only required claims which positively affects performance.
+    // +optional
+    ClaimsFilter []string
+}
+```
+
+1. `Issuer` - is a sections for external provider specific settings, e.g., OIDC discovery URL. 
+  There is also default validation settings (ClientID, SkipOIDCValidations) that's available out of the box for flag-based 
+  OIDC provider, but for structured config they can be disabled.
+
+    ```go
+    type Issuer struct {
+        // URL points to the issuer URL in a format schema://url/path.
+        URL string
+        // CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+        CertificateAuthorityData []byte
+        // ClientID the JWT must be issued for, the "sub" field. This plugin only trusts a single
+        // client to ensure the plugin can be used with public providers.
+        // Do not affect anything with the SkipOIDCValidations option enabled.
+        // +optional
+        ClientID string
+        // SkipOIDCValidations is a flag to turn off issuer validation, client id validation.
+        // OIDC related checks.
+        // +optional
+        SkipOIDCValidations bool
+   }
+   ```
+
+2. `ClaimValidationRules` - additional rules for authorization.
+    ```go
+    type ClaimValidationRule struct {
+        // Rule is a logical expression that is written in CEL https://github.com/google/cel-go.
+        Rule string
+        // Message customize returning message for validation error of the particular rule.
+        // +optional
+        Message string
+    }
+    ```
+
+    For validation expressions, the CEL is used. They are similar to validations functions for [Custom Resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions).
+    `Rule` expression should always evaluate a boolean. Token `claims` are passed to CEL expressions as a dynamic map `decls.NewMapType(decls.String, decls.Dyn)`.
+
+    You can find a snippet of validation rules below:
+
+    ```yaml
+    claimValidationRules:
+    
+    - rule: 'claims.aud == "charmander" || claims.aud == "bulbasaur"'
+      message: clients other than charmander or bulbasaur are not allowed
+    
+    - rule: 'claims.roles.exists(r, r == "kubernetes-user")'
+      message: only kubernetes-user group members can access the cluster
```

Comment:
This feels like it could be more of an authorization rule. Let's think at some point about how we document this, so people don't put more complexity into `claimValidationRules` than is wise.

I think I'd prefer to have rule that only accepts current tokens where the total lifetime (`exp - nbf`) is less than 24 hours, because that's more of an authentication thing.

Reply:
It seems reasonable. I will pick your variant.